[core] Add monitoring in raylet for resouce view

ZacAttack opened the pull request:
## Description

We today have very little observability into pubsub.  On a raylet one of the most important states that need to be propagated through the cluster via pubsub is cluster membership.  All raylets should in an eventual BUT timely fashion agree on the list of available nodes.  This metric just emits a simple counter to keep track of the node count.

More pubsub observability to come.

## Related issues
> Link related issues: "Fixes #1234", "Closes #1234", or "Related to #1234".

## Additional information
> Optional: Add implementation details, API changes, usage examples, screenshots, etc.

ZacAttack committed:
Add metrics

Signed-off-by: zac <zac@anyscale.com>

ZacAttack committed:
Merge branch 'master' into scheduler-hax

Signed-off-by: zac <zac@anyscale.com>

edoakes commented on src/ray/stats/metric_defs.cc:
```diff
@@ -124,6 +124,11 @@ DEFINE_stats(scheduler_placement_time_s,
+             "Number of nodes tracked in the ClusterResourceManager.",
+             (),
+             (),
+             ray::stats::GAUGE);
```
feels a little too tightly coupled to implementation details

maybe something like:
```suggestion
DEFINE_stats(num_known_nodes_in_cluster,
             "Number of nodes tracked in the ClusterResourceManager.",
             (),
             (),
             ray::stats::GAUGE);
```

edoakes replied:
also -- I think by prometheus best practices, this metric should not include "count" in the name because it's a gauge (which means it does not monotonically increase)

edoakes reviewed (commented):
What other metrics do you expect to add? One that would be useful would be number of nodes that are fully at capacity (or conversely, number of nodes with schedulable slots)

ZacAttack committed:
Update src/ray/stats/metric_defs.cc

Co-authored-by: Edward Oakes <ed.nmi.oakes@gmail.com>
Signed-off-by: Zac Policzer <zacattackftw@gmail.com>

ZacAttack committed:
update name from comment

Signed-off-by: zac <zac@anyscale.com>

can-anyscale reviewed (approved):
looking good, i'm migrating away from metric_defs.h/cc but have not touched raylet yet

can-anyscale commented on src/ray/stats/metric_defs.cc:
```diff
@@ -124,6 +124,11 @@ DEFINE_stats(scheduler_placement_time_s,
              ({0.1, 1, 10, 100, 1000, 10000}, ),
              ray::stats::HISTOGRAM);
+DEFINE_stats(local_resource_view_node_count,
+             "Number of nodes tracked locally by the reporting node.",
```
This description is a bit difficult to understand though, maybe clarify it by adding “as opposed to GCS” to show what it contrasts with.

For example: “Number of nodes tracked locally by the raylet (as opposed to GCS).”

ZacAttack committed:
[core][build] Update ray to build with python 3.14 and add build images

This contains necessary updates for the build to pass and for CI to
create 3.14 images.  IT DOES NOT guarantee that all features of 3.14
work with ray.

ZacAttack committed:
Revert "[core][build] Update ray to build with python 3.14 and add build images"

This reverts commit 2da1ce513bba005b85293bf32d78bd3a45af9f36.

aslonnie reviewed (commented):
does not seem need CI team review?

ZacAttack committed:
Merge branch 'master' into scheduler-hax

Signed-off-by: zac <zac@anyscale.com>
